src/graph.rs

```diff
@@ -22,11 +24,23 @@ impl<'hw> ArrayPlaceholder<'hw> {
     /// A reference to the inner `Shape` object.
     pub(crate) fn shape(&self) -> &Shape {
         match self {
-            Self::Unassigned(ref shape) => shape,
+            Self::Unassigned(ref shape, _) => shape,
             Self::Assigned(ref array) => array.shape(),
         }
     }
 
+    /// Obtains the `Hardware` of this placeholder.
+    ///
+    /// # Returns
+    ///
+    /// A reference to the `Hardware` object.
+    pub(crate) fn hardware(&self) -> &'hw RefCell<dyn Hardware> {
+        match self {
+            Self::Unassigned(_, hardware) => hardware,
+            Self::Assigned(ref array) => array.hardware(),
```

Comment:
Is `ref` necessary? It is used in many places, but I think most of them are unnecessary.
```suggestion
            Self::Assigned(array) => array.hardware(),
```

Reply:
Yes in some cases the destructured thing is already a reference, so it is unnecessary.

I guess there is some ambiguity of the language itself in the context of destructuring: 🤔 

<img width="200" alt="Untitled" src="https://user-images.githubusercontent.com/1023695/155742811-1710f35b-541a-4b0c-8b98-e807f365091e.png">

Reply:
fixed fc1fd0a

Reply:
Oh the ambiguity is explicitly introduced. sigh,
https://github.com/rust-lang/rfcs/blob/master/text/2005-match-ergonomics.md